src/Event.cs

```diff
@@ -62,41 +62,118 @@ public SubscriptionBuilder<EventContext> OnError(Callback callback)
             return this;
         }
 
-        public SubscriptionHandle<EventContext> Subscribe(params string[] querySqls) => new(conn, Applied, Error, querySqls);
+        public SubscriptionHandle<EventContext> Subscribe(string querySql) => new(conn, Applied, Error, querySql);
 
         public void SubscribeToAllTables()
         {
-            Subscribe("SELECT * FROM *");
+            // make sure we use the legacy handle constructor here, even though there's only 1 query.
+            new SubscriptionHandle<EventContext>(conn, Applied, Error, new string[] { "SELECT * FROM *" });
         }
     }
 
     public interface ISubscriptionHandle
     {
         void OnApplied(IEventContext ctx);
+        void OnError(IEventContext ctx);
+        void OnEnded(IEventContext ctx);
     }
 
     public class SubscriptionHandle<EventContext> : ISubscriptionHandle
         where EventContext : IEventContext
     {
         private readonly SubscriptionBuilder<EventContext>.Callback? onApplied;
+        private readonly SubscriptionBuilder<EventContext>.Callback? onError;
+        private SubscriptionBuilder<EventContext>.Callback? onEnded;
+
+        public bool IsEnded
+        {
+            get; private set;
+        }
+
+        public bool IsActive
+        {
+            get; private set;
+        }
 
         void ISubscriptionHandle.OnApplied(IEventContext ctx)
         {
             IsActive = true;
             onApplied?.Invoke((EventContext)ctx);
         }
 
+        void ISubscriptionHandle.OnEnded(IEventContext ctx)
+        {
+            IsActive = false;
+            IsEnded = true;
+            onEnded?.Invoke((EventContext)ctx);
+        }
+
+        void ISubscriptionHandle.OnError(IEventContext ctx)
+        {
+            IsActive = false;
+            IsEnded = true;
+            onError?.Invoke((EventContext)ctx);
+        }
+
+        /// <summary>
+        /// TODO: remove this constructor once legacy subscriptions are removed.
+        /// </summary>
+        /// <param name="conn"></param>
+        /// <param name="onApplied"></param>
+        /// <param name="onError"></param>
+        /// <param name="querySqls"></param>
         internal SubscriptionHandle(IDbConnection conn, SubscriptionBuilder<EventContext>.Callback? onApplied, SubscriptionBuilder<EventContext>.Callback? onError, string[] querySqls)
         {
+            IsActive = false;
+            IsEnded = false;
+            this.onApplied = onApplied;
```

Comment:
If this constructor doesn't assign `onError`, should the param be removed maybe?